### ISSUE TITLE
Event dispatch retries forever when event observer shutdown first

### ISSUE THREAD

**Issue by zone117x** [stacks-network/stacks-blockchain#2575]
Another issue I've noticed is that if the event observer (API) is shutdown first, the stacks-node will get stuck in a event dispatch retry loop and never shutdown. This message repeating:
```
Event dispatcher: connection failed  - Os { code: 61, kind: ConnectionRefused, message: "Connection refused" }
```

This may be difficult for infrastructure that can't always control the timing of which services are restarted first.

_Originally posted by @zone117x in https://github.com/blockstack/stacks-blockchain/issues/2547#issuecomment-809508925_

**Comment by zone117x** [comment]
I think this is simply a matter of adding `should_keep_running: Option<Arc<AtomicBool>>` to the event dispatcher loop.

**Comment by lgalabru** [comment]
To be addressed with https://github.com/blockstack/stacks-blockchain/pull/2570.
That being said, we should do our best to make sure that the event observers are only being terminated **once** the stacks-node exited, because the stacks-node will **not** try to re-send the lost messages on next relaunch.

**Comment by zone117x** [comment]
@lgalabru that would make it unsafe for the API to be able to restart without also restarting the stacks-node. Also, while I'm sure we can force this behavior on our infra, it would be difficult to enforce 3rd parties to implement this brittle setup correctly. Is there anything we can do to prevent this?

**Comment by jcnelson** [comment]
This fixed?

**Comment by jcnelson** [comment]
Closing due to inactivity.  I think this is fixed, but please reopen if not.

**Comment by planbetterHQ** [comment]
We've just experienced this issue upgrading the nodes and the api.

**Comment by planbetterHQ** [comment]
A restart of the Stacks node unblocks the situation.
